Duplicate multibinding source errors when using createDynamicGraphFactory intrinsic

### Description

If you use `createDynamicGraphFactory` in an `androidTest` and your main graph contains multibindings, you end up with "Duplicate multibinding source" errors.

E.g.: 

In app:

```kotlin
interface Listener

@DependencyGraph(
    scope = AppScope::class,
    bindingContainers = [AppModule::class],
)
interface AppGraph {
    @DependencyGraph.Factory
    interface Factory {
        fun create(): AppGraph
    }

    @Multibinds(allowEmpty = true)
    fun listeners(): Set<Listener>
}

@BindingContainer
interface AppModule {
    companion object {
        @Provides
        @IntoSet
        fun provideListener(): Listener = object : Listener {}
    }
}
```

In androidTest:

```kotlin
class TestApp {
    fun createGraph(): AppGraph {
        return createDynamicGraphFactory<AppGraph.Factory>(
            TestOverrideModule,
        ).create()
    }
}

@BindingContainer
object TestOverrideModule
```

### Self-contained Reproducer

I've attached a sample project for repro:
```bash
$ ./gradlew :app:compileDebugAndroidTestKotlin // fails
Caused by: java.lang.IllegalStateException: Duplicate multibinding source: @dev.zacsweers.metro.internal.MultibindingElement("com.example.metrorepo.Listener", "335189748") com.example.metrorepo.Listener. This is possibly a bug in the Metro compiler, please report it with details and/or a reproducer to https://github.com/zacsweers/metro. 
        at dev.zacsweers.metro.compiler.ir.graph.IrBinding$Multibinding.addSourceBinding(IrBinding.kt:981)
        at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.registerMultibindingContribution(BindingLookup.kt:307)
        at dev.zacsweers.metro.compiler.ir.graph.BindingLookup.putBinding(BindingLookup.kt:190)
        at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate$putBinding(BindingGraphGenerator.kt:97)
        at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.access$generate$putBinding(BindingGraphGenerator.kt:46)
        at dev.zacsweers.metro.compiler.ir.graph.BindingGraphGenerator.generate(BindingGraphGenerator.kt:231)
        at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.validateDependencyGraph(DependencyGraphTransformer.kt:242)
        at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.access$validateDependencyGraph(DependencyGraphTransformer.kt:106)
        at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraphInner(DependencyGraphTransformer.kt:170)
        at dev.zacsweers.metro.compiler.ir.transformers.DependencyGraphTransformer.processGraph$compiler(DependencyGraphTransformer.kt:130)
        at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generateInner(MetroIrGenerationExtension.kt:189)
        at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generate(MetroIrGenerationExtension.kt:70)
        at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:571)
        ... 43 more
```

But compiling the app sources succeeds:
```
$ ./gradlew :app:assembleDebug 
BUILD SUCCESSFUL in 364ms
```

[metro-multibinding-repro.zip](https://github.com/user-attachments/files/26061472/metro-multibinding-repro.zip)

### Metro version

0.11.4

### Previous working version

N/A

### IDE version

_No response_

### Context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Changelog
 ### Fixes
 
 - **[IR]**: Fix `IllegalArgumentException` thrown when there are multiple top-level functions with the same name but only one is annotated with `@Inject`.
+- **[IR]**: Only store a given binding container's own provider factories in metro metadata. This resolves a bug where we could end up duplicate-processing upstream providers in dynamic factories.
 
 ### Contributors
 

--- a/compiler-tests/src/test/data/box/dependencygraph/dynamic/MultibindsDoNotReportAsDupes.kt
+++ b/compiler-tests/src/test/data/box/dependencygraph/dynamic/MultibindsDoNotReportAsDupes.kt
@@ -1,0 +1,30 @@
+// Regression test to ensure we only process upstream binding containers once
+// https://github.com/ZacSweers/metro/issues/2005
+// MODULE: lib
+interface Listener
+
+@DependencyGraph(scope = AppScope::class, bindingContainers = [AppModule::class])
+interface AppGraph {
+  @DependencyGraph.Factory
+  interface Factory {
+    fun create(): AppGraph
+  }
+
+  @Multibinds(allowEmpty = true) fun listeners(): Set<Listener>
+}
+
+@BindingContainer
+interface AppModule {
+  companion object {
+    @Provides @IntoSet fun provideListener(): Listener = object : Listener {}
+  }
+}
+
+// MODULE: main(lib)
+@BindingContainer object TestOverrideModule
+
+fun box(): String {
+  val graph = createDynamicGraphFactory<AppGraph.Factory>(TestOverrideModule).create()
+  assertNotNull(graph.listeners())
+  return "OK"
+}

--- a/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
+++ b/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
@@ -962,6 +962,12 @@ public class BoxTestGenerated extends AbstractBoxTest {
       }
 
       @Test
+      @TestMetadata("MultibindsDoNotReportAsDupes.kt")
+      public void testMultibindsDoNotReportAsDupes() {
+        runTest("compiler-tests/src/test/data/box/dependencygraph/dynamic/MultibindsDoNotReportAsDupes.kt");
+      }
+
+      @Test
       @TestMetadata("ReplacingBindingContainers.kt")
       public void testReplacingBindingContainers() {
         runTest("compiler-tests/src/test/data/box/dependencygraph/dynamic/ReplacingBindingContainers.kt");

--- a/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/FastInitBoxTestGenerated.java
+++ b/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/FastInitBoxTestGenerated.java
@@ -962,6 +962,12 @@ public class FastInitBoxTestGenerated extends AbstractFastInitBoxTest {
       }
 
       @Test
+      @TestMetadata("MultibindsDoNotReportAsDupes.kt")
+      public void testMultibindsDoNotReportAsDupes() {
+        runTest("compiler-tests/src/test/data/box/dependencygraph/dynamic/MultibindsDoNotReportAsDupes.kt");
+      }
+
+      @Test
       @TestMetadata("ReplacingBindingContainers.kt")
       public void testReplacingBindingContainers() {
         runTest("compiler-tests/src/test/data/box/dependencygraph/dynamic/ReplacingBindingContainers.kt");

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/IrGraphGenerator.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/graph/IrGraphGenerator.kt
@@ -311,7 +311,17 @@ internal class IrGraphGenerator(
       if (!graphClass.origin.isSyntheticGeneratedGraph) {
         trace("Generate Metro metadata") {
           // Finally, generate metadata
-          val graphProto = node.toProto(bindingGraph = bindingGraph)
+          // Use only the graph's own provider factories (not those from binding containers)
+          // for metadata. Binding container factories are resolved independently by consumers.
+          val ownProviderFactories =
+            metroDeclarations
+              .findBindingContainer(node.sourceGraph)
+              ?.providerFactories
+              ?.values
+              .orEmpty()
+              .toSet()
+          val graphProto =
+            node.toProto(bindingGraph = bindingGraph, ownProviderFactories = ownProviderFactories)
           graphMetadataReporter.write(node, bindingGraph)
           val metroMetadata = createMetroMetadata(dependency_graph = graphProto)
 

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/metadata.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/metadata.kt
@@ -76,7 +76,10 @@ internal var IrClass.metroMetadata: MetroMetadata?
     context.metadataDeclarationRegistrar.addCustomMetadataExtension(this, PLUGIN_ID, value.encode())
   }
 
-internal fun GraphNode.toProto(bindingGraph: IrBindingGraph): DependencyGraphProto {
+internal fun GraphNode.toProto(
+  bindingGraph: IrBindingGraph,
+  ownProviderFactories: Set<ProviderFactory>,
+): DependencyGraphProto {
   var multibindingAccessors = BitField()
   val accessorNames =
     accessors
@@ -92,7 +95,7 @@ internal fun GraphNode.toProto(bindingGraph: IrBindingGraph): DependencyGraphPro
 
   return createGraphProto(
     isGraph = true,
-    providerFactories = providerFactories.values.flatten(),
+    providerFactories = ownProviderFactories,
     accessorNames = accessorNames,
     multibindingAccessorIndices = multibindingAccessors.toIntList(),
   )

